builtin/logical/database/backend.go

```diff
@@ -286,3 +485,329 @@ cassandra, mssql, mysql, postgres
+		}
+		if role == nil || role.StaticAccount == nil {
+			continue
+		}
+
+		if err := b.pushItem(&queue.Item{
+			Key:      roleName,
+			Priority: role.StaticAccount.LastVaultRotation.Add(role.StaticAccount.RotationPeriod).Unix(),
+		}); err != nil {
+			log.Warn("unable to enqueue item", "error", err, "role", roleName)
+		}
+	}
+}
+
+// runTicker kicks off a periodic ticker that invoke the automatic credential
+// rotation method at a determined interval. The default interval is 5 seconds.
+func (b *databaseBackend) runTicker(ctx context.Context, s logical.Storage) {
+	b.logger.Info("starting periodic ticker")
+	tick := time.NewTicker(queueTickInterval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+			b.rotateCredentials(ctx, s)
+
+		case <-ctx.Done():
+			b.logger.Info("stopping periodic ticker")
+			return
+		}
+	}
+}
+
+// setCredentialsWAL is used to store information in a WAL that can retry a
+// credential setting or rotation in the event of partial failure.
+type setCredentialsWAL struct {
+	NewPassword string
+	OldPassword string
+	RoleName    string
+	Username    string
+
+	LastVaultRotation time.Time
+	Statements        dbplugin.Statements
+}
+
+// rotateCredentials sets a new password for a static account. This method is
+// invoked in the runTicker method, which is in it's own go-routine, and invoked
+// periodically (approximately every 5 seconds).
+//
+// This method loops through the priority queue, popping the highest priority
+// item until it encounters the first item that does not yet need rotation,
+// based on the current time.
+func (b *databaseBackend) rotateCredentials(ctx context.Context, s logical.Storage) error {
+	for {
+		item, err := b.credRotationQueue.Pop()
+		if err != nil {
+			if err == queue.ErrEmpty {
+				return nil
+			}
+			return err
+		}
+
+		// validate the role still exists
+		role, err := b.StaticRole(ctx, s, item.Key)
+		if err != nil {
+			b.logger.Warn("unable load role", "role", item.Key, "error", err)
+			continue
+		}
+		if role == nil {
+			b.logger.Warn("role not found", "role", item.Key, "error", err)
+			continue
+		}
+
+		if time.Now().Unix() > item.Priority {
+			// We've found our first item not in need of rotation
+			input := &setStaticAccountInput{
+				RoleName: item.Key,
+				Role:     role,
+			}
+
+			// If there is a WAL entry related to this Role, the corresponding WAL ID
+			// should be stored in the Item's Value field.
+			if walID, ok := item.Value.(string); ok {
+				walEntry := b.findStaticWAL(ctx, s, walID)
+				if walEntry != nil && walEntry.NewPassword != "" {
+					input.Password = walEntry.NewPassword
+					input.WALID = walID
+				}
+			}
+
+			resp, err := b.setStaticAccount(ctx, s, input)
+			if err != nil {
+				b.logger.Warn("unable to rotate credentials in periodic function", "error", err)
+				// update the priority to re-try this rotation and re-add the item to
+				// the queue
+				item.Priority = time.Now().Add(10 * time.Second).Unix()
```

Comment:
Why do we update the priority on a failed rotation?  Is it intended to be a backoff of sorts?

Reply:
Yes. I changed the wording of the comment above to make that more clear.